protons/app/proposals.py

```diff
@@ -159,3 +160,296 @@ def propose_states(self, drive, residue_pool_indices):
+        list(tuple(int,int),) -  list of tuples with (initial, final) states of the water molecule/ion as int, one tuple for each requested swap
+            By saltswap convention
+             - 0 is water
+             - 1 is cation
+             - 2 is anion
+        float - log (probability of reverse proposal)/(probability of forward proposal)
+        """
+
+        return list(), list(), float()
+
+
+class UniformSwapProposal(_SaltSwapProposal):
+    """Uniformly selects as many water/ions as needed from the array of available waters/ions."""
+
+    def __init__(self, cation_coefficient: float = 0.5):
+        """Instantiate a UniformSwapProposal.
+
+        Parameters
+        ----------
+        cation_coefficient - float, optional. Must be between 0 and 1, default 0.5.
+            The fraction of the chemical potential of a water pair -> salt pair transformation that is attributed to the
+            cation -> water transformation.
+        """
+        if not 0.0 <= cation_coefficient <= 1.0:
+            raise ValueError("The cation coefficient should be between 0 and 1.")
+
+        self._cation_weight = cation_coefficient
+        self._anion_weight = 1.0 - cation_coefficient
+
+    @staticmethod
+    def _select_ion_water_swaps(drive, charge_to_counter):
+        """Select ions or water molecule swap procedure to facilitate maintaining charge neutrality while changing protonation states.
+
+        Notes
+        -----
+        Technically, for charge differences of 2 we could change one anion into a cation or vice versa.
+        This code would instead remove one anion, and change one water into a cation.
+
+        Parameters
+        ----------
+        drive - a ProtonDrive object
+        charge_to_counter - the total difference in charge that needs to be countered.
+
+        Returns
+        -------
+        dict(water_to_cation, water_to_anion, cation_to_water, anion_to_water)
+
+        """
+
+        # Note that we don't allow for direct transitions between ions of different charge.
+        swaps = dict(water_to_cation=0, water_to_anion=0, cation_to_water=0, anion_to_water=0)
+        excess_ions = int(drive.excess_ions)  # copy
+
+        while abs(charge_to_counter) > 0:
+
+            # A net amount of cations were previously added to the system
+            if excess_ions > 0:
+                charge_to_counter, excess_ions = UniformSwapProposal._swap_excess_cations(charge_to_counter, excess_ions, swaps)
+
+            # No additional ions were previously added by the swapper.
+            elif excess_ions == 0:
+                charge_to_counter, excess_ions = UniformSwapProposal._swap_no_excess_ions(charge_to_counter, excess_ions, swaps)
+
+            # A net amount of anions were previously added to the system
+            elif excess_ions < 0:
+                charge_to_counter, excess_ions = UniformSwapProposal._swap_excess_anions(charge_to_counter, excess_ions, swaps)
+
+        return swaps
```

Comment:
I've double checked this and these moves appear correctly reversible. 

Although the above looks fine, I'm curious about the reasons behind your selection logic. In the scheme above, if the system has an excess of anions (`excess_ions`< 0) and `charge_to_counter` > 0, your scheme will add another anion; wouldn't it be more intuitive to add a cation instead? Or is that more complicated to make it reversible? One could also use probabilistic selection here instead, where the choice to insert or delete an anion/cation is based on their relative quantities in addition to the charge change.

Reply:
>if the system has an excess of anions (excess_ions< 0) and charge_to_counter > 0, your scheme will add another anion; wouldn't it be more intuitive to add a cation instead?

I assume you mean remove cation. Adding a cation would be the wrong charge direction.

Just to summarize the scheme:

For one single charge change, 
if the direction of the titration state change is positive (e.g. `NH0` -> `NH+`, or `O-` -> `OH`),
-   if the excess ions are anions: remove water, add anions
-   if the excess ions are cations: remove cations, add water.

If the direction of the titration state change is negative (e.g. `NH+` -> `NH0)`, or `OH` -> `O-`)
-  if the excess ions are anions: remove anions, add water
-  if the excess ions are cations: remove water, add cations
  
My thoughts behind this logic were mostly practical.

* From a practical point, this makes sure that I dont add an ion first, and on the reverse protocol add an ion of the opposite charge, and end up with potential accumulation of salt.
* Presumably the water reservoir will be larger than the ion reservoir. I therefore only want to add ions if possible, not remove them. Otherwise, potentially for a large protein we would sooner start depleting ions leading to impossible moves eventually. If I clearly have an excess, there is a guarantee that an ion is available. Maybe there are other ways around this.

I'd be happy to use a different scheme if you have suggestions. If you see a problem with the method, or my reasoning, please let me know and we can work on an alternative. If at any point in time we want to change the way it works, all one has to do is implement a new SaltSwapProposal subclass and use that instead.